executor/brie.go

```diff
@@ -461,8 +846,93 @@ func (gs *tidbGlueSession) StartProgress(ctx context.Context, cmdName string, to
 func (gs *tidbGlueSession) Record(name string, value uint64) {
 	switch name {
 	case "BackupTS":
-		gs.info.backupTS = value
+		gs.info.setBackupTS(value)
 	case "Size":
-		gs.info.archiveSize = value
+		gs.info.setArchiveSize(value)
+	// importglue.RecordEstimatedChunk is sent only once and before importglue.RecordFinishedChunk
+	case importglue.RecordEstimatedChunk:
+		gs.StartProgress(context.Background(), cmdImport, int64(value), false)
+	case importglue.RecordFinishedChunk:
+		current := int64(value)
+		gs.progress.lock.Lock()
+		if current >= gs.progress.total {
+			// TODO(lance6717): to be more exact, lightning could report status of restoring tables instead of chunks
+			gs.progress.setCmd(cmdPostProcess)
+		}
+		atomic.StoreInt64(&gs.progress.current, current)
+		gs.progress.lock.Unlock()
+	}
+}
+
+func (gs *tidbGlueSession) Flush(ctx context.Context, taskID uint64) {
+	sql := fmt.Sprintf(`UPDATE mysql.brie_tasks SET
+					exec_time = '%s',
+					finish_time = '%s',
+					ts = %d,
+					data_size = %d,
+					status = '%s',
+					progress = %f,
+					message = '%s'
+				WHERE id = %d`,
+		gs.info.getExecTime().String(),
+		gs.info.getFinishTime().String(),
+		gs.info.getBackupTS(),
+		gs.info.getArchiveSize(),
+		gs.progress.getCmd(),
+		gs.progress.getFraction(),
+		base64.StdEncoding.EncodeToString([]byte(gs.info.getMessage())),
+		taskID)
+
+	sqlExecutor := gs.se.(sqlexec.RestrictedSQLExecutor)
+	_, _, err := sqlExecutor.ExecRestrictedSQLWithContext(ctx, sql)
+	if err != nil {
+		logutil.Logger(ctx).Error("failed to update BRIE task table",
+			zap.Uint64("taskID", taskID),
+			zap.Error(err))
+	}
+}
+
+func (gs *tidbGlueSession) OwnsSQLExecutor() bool {
+	return false
+}
+
+func (gs *tidbGlueSession) GetSQLExecutor() importglue.SQLExecutor {
+	return gs
+}
+
+func (gs *tidbGlueSession) GetParser() *parser.Parser {
+	p := parser.New()
+	p.SetSQLMode(gs.se.GetSessionVars().SQLMode)
+	return p
+}
+
+func (gs *tidbGlueSession) GetDB() (*sql.DB, error) {
+	return nil, errors.New("tidbGlueSession can't perform GetDB")
+}
+
+func (gs *tidbGlueSession) GetTables(ctx context.Context, schemaName string) ([]*model.TableInfo, error) {
+	is := domain.GetDomain(gs.se).InfoSchema()
+	tables := is.SchemaTables(model.NewCIStr(schemaName))
+	tbsInfo := make([]*model.TableInfo, len(tables))
+	for i := range tbsInfo {
+		tbsInfo[i] = tables[i].Meta()
+	}
+	return tbsInfo, nil
+}
+
+func (gs *tidbGlueSession) GetSession() (checkpoints.Session, error) {
+	s, err := CreateSessionForBRIEFunc(gs.se)
```

Comment:
Can we return `gs.se` directly?

Reply:
lightning will use `session.Execute` concurrently, so we can't return one session to them all.

and some SQL are seperately executed such as "BEGIN", some update, `session.CommitTxn`, thus can't use `RestrictedSQLExecutor`